Bato.to k07 server causes read/download failure

### Source information

Bato.to 1.4.58

### Source language

All

### Steps to reproduce

1. Download/read a chapter
2. If an image source originates from k07 server, the chapter fails to download / image fails to load

### Expected behavior

Should correctly fallback to n server

### Actual behavior

Download errors out / image fails to load

### Mihon/Tachiyomi version

0.19.3

### Android version

Android 14

### Other details

Recently, unlike other k servers that return http errors such as 404/503, k07 server began returning `HTTP/1.1 301 Moved Permanently`, and redirects to `plateforme-cabinet-guerin.fr` domain. For example, this [k07 link](https://k07.mbhiz.org/media/mbch/4e0/695aefb5c3860c9edf8b30e4/134930731_1280_2400_372080.webp) would end up at this [plateforme-cabinet-guerin.fr](https://plateforme-cabinet-guerin.fr/media/mbch/4e0/695aefb5c3860c9edf8b30e4/134930731_1280_2400_372080.webp) link. My guess is that this redirection behaviour causes the extension to fail to load images that are on the k07 server, and also the cause of a previous closed ticket 3 days ago #12509 . Perhaps the fallback logic is currently too strict and requires specifically 404 or 503 to trigger?

This issue also affects the v4 extension.

I have also replicated this issue on:
Mihon 0.19.3 (Android 14)
Tachimanga v4.11 (iOS 26.1 & iOS 18.7)
Tachidesk (Suwayomi-Server v2.1.1867, Windows)

Interestingly, in the Mihon app, if i continuously restart the failed download, it eventually succeeds, which is why I assume this issue hasn't been reported. However, Tachimanga and Suwayomi-Server will always fail.

Here is a log snippet from Tachidesk. It would repeat a few times trying to download this image, then error out.
```
2026-01-05 20:59:23.540 [OkHttp https://n10.mbopg.org/...] DEBUG eu.kanade.tachiyomi.network.NetworkHelper - --> GET https://n10.mbopg.org/media/mbch/4e0/695aefb5c3860c9edf8b30e4/134930732_1280_2400_342028.webp h2
2026-01-05 20:59:23.850 [OkHttp https://n10.mbopg.org/...] DEBUG eu.kanade.tachiyomi.network.NetworkHelper - <-- 200 https://n10.mbopg.org/media/mbch/4e0/695aefb5c3860c9edf8b30e4/134930732_1280_2400_342028.webp (310ms, 342028-byte body)
2026-01-05 20:59:24.455 [OkHttp https://k07.mbhiz.org/...] DEBUG eu.kanade.tachiyomi.network.NetworkHelper - --> GET https://k07.mbhiz.org/media/mbch/4e0/695aefb5c3860c9edf8b30e4/134930731_1280_2400_372080.webp h2
2026-01-05 20:59:24.779 [OkHttp https://k07.mbhiz.org/...] DEBUG eu.kanade.tachiyomi.network.NetworkHelper - <-- 301 https://k07.mbhiz.org/media/mbch/4e0/695aefb5c3860c9edf8b30e4/134930731_1280_2400_372080.webp (323ms, unknown-length body)
2026-01-05 20:59:45.804 [DefaultDispatcher-worker-11] WARN  suwayomi.tachidesk.manga.impl.download.Downloader source(7890050626002177109) - downloadChapter(Miss Not-So Sidekick [Official] (757) - Side Story 2 Episode 13 (45146) | state= Queued, tries= 0, progress= 0.0)) - failed due to
java.io.IOException: Failed to connect to plateforme-cabinet-guerin.fr/51.159.97.100:443
	at eu.kanade.tachiyomi.source.online.HttpSource.getImage$suspendImpl(HttpSource.kt:374)
	at eu.kanade.tachiyomi.source.online.HttpSource.getImage(HttpSource.kt)
	at suwayomi.tachidesk.manga.impl.Page$getPageImage$6.invokeSuspend(Page.kt:129)
	at suwayomi.tachidesk.manga.impl.Page$getPageImage$6.invoke(Page.kt)
	at suwayomi.tachidesk.manga.impl.Page$getPageImage$6.invoke(Page.kt)
	at suwayomi.tachidesk.manga.impl.util.storage.ImageResponse.getImageResponse(ImageResponse.kt:68)
	at suwayomi.tachidesk.manga.impl.Page.getPageImage(Page.kt:128)
	at suwayomi.tachidesk.manga.impl.Page.getPageImage$default(Page.kt:49)
	at suwayomi.tachidesk.manga.impl.download.fileProvider.ChaptersFilesProvider.downloadImpl(ChaptersFilesProvider.kt:141)
	at suwayomi.tachidesk.manga.impl.download.fileProvider.ChaptersFilesProvider.access$downloadImpl(ChaptersFilesProvider.kt:68)
	at suwayomi.tachidesk.manga.impl.download.fileProvider.ChaptersFilesProvider$downloadImpl$1.invokeSuspend(ChaptersFilesProvider.kt)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:100)
	at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:586)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:829)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:717)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:704)
Caused by: java.net.ConnectException: Failed to connect to plateforme-cabinet-guerin.fr/51.159.97.100:443
	at okhttp3.internal.connection.ConnectPlan.connectSocket(ConnectPlan.kt:282)
	at okhttp3.internal.connection.ConnectPlan.connectTcp(ConnectPlan.kt:140)
	at okhttp3.internal.connection.FastFallbackExchangeFinder$launchTcpConnect$1.runOnce(FastFallbackExchangeFinder.kt:141)
	at okhttp3.internal.concurrent.TaskRunner$runnable$1.run(TaskRunner.kt:81)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
	at java.base/java.lang.Thread.run(Unknown Source)
Caused by: java.net.ConnectException: Connection timed out: getsockopt
	at java.base/sun.nio.ch.Net.pollConnect(Native Method)
	at java.base/sun.nio.ch.Net.pollConnectNow(Unknown Source)
	at java.base/sun.nio.ch.NioSocketImpl.timedFinishConnect(Unknown Source)
	at java.base/sun.nio.ch.NioSocketImpl.connect(Unknown Source)
	at java.base/java.net.SocksSocketImpl.connect(Unknown Source)
	at java.base/java.net.Socket.connect(Unknown Source)
	at okhttp3.internal.platform.Platform.connectSocket(Platform.kt:147)
	at okhttp3.internal.connection.ConnectPlan.connectSocket(ConnectPlan.kt:280)
	... 6 common frames omitted
```

### Acknowledgements

- [x] I have searched the existing issues and this is a new ticket, **NOT** a duplicate or related to another open or closed issue.
- [x] I have written a short but informative title.
- [x] I have updated the app to version **[0.15.3](https://github.com/mihonapp/mihon/releases/latest)**.
- [x] I have updated all installed extensions.
- [x] I have tried the [troubleshooting guide](https://mihon.app/docs/guides/troubleshooting/).
- [x] If this is an issue with the app itself, I should be opening an issue in the app repository.
- [x] I will fill out all of the requested information in this form.

### <!-- footer -->

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/keiyoushi/extensions-source/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

--- a/src/all/batoto/build.gradle
+++ b/src/all/batoto/build.gradle
@@ -1,7 +1,7 @@
 ext {
     extName = 'Bato.to'
     extClass = '.BatoToFactory'
-    extVersionCode = 59
+    extVersionCode = 60
     isNsfw = true
 }
 

--- a/src/all/batoto/src/eu/kanade/tachiyomi/extension/all/batoto/BatoTo.kt
+++ b/src/all/batoto/src/eu/kanade/tachiyomi/extension/all/batoto/BatoTo.kt
@@ -16,6 +16,7 @@ import keiyoushi.utils.getPreferencesLazy
 import okhttp3.HttpUrl.Companion.toHttpUrl
 import okhttp3.Response
 import rx.Observable
+import java.util.concurrent.ConcurrentHashMap
 
 open class BatoTo(
     final override val lang: String,
@@ -191,5 +192,55 @@ open class BatoTo(
         )
 
         private val mirrors = mirrorsV2 + mirrorsV4
+    }
+}
+
+/**
+ * Manages image server fallback logic, including blacklisting and backoff tracking.
+ */
+class ImageServerManager() {
+    val serverPattern = Regex("https://([a-zA-Z]\\d{2})")
+
+    val fallbackServers = listOf(
+        "n03", "n00", "n01", "n02", "n04", "n05", "n06", "n07", "n08", "n09", "n10",
+        "k03", "k06", "k00", "k01", "k02", "k04", "k05", "k08", "k09",
+    )
+    val blacklist = listOf("k07")
+
+    // Server status tracking
+    data class ServerStatus(
+        val canBackoff: Boolean,
+        val statusCode: Int = 0,
+        val timestamp: Long = System.currentTimeMillis(),
+    )
+
+    val serverStatus = ConcurrentHashMap<String, ServerStatus>()
+
+    val BACKOFF_DURATION_MS = 3_600_000L // 1 hour
+
+    fun shouldSkip(server: String): Boolean {
+        return server in blacklist || isInBackoff(server)
+    }
+
+    fun isInBackoff(server: String): Boolean {
+        val status = serverStatus[server] ?: return false
+        return status.canBackoff && System.currentTimeMillis() - status.timestamp < BACKOFF_DURATION_MS
+    }
+
+    fun recordImageServerStatus(server: String, statusCode: Int) {
+        val now = System.currentTimeMillis()
+        serverStatus[server] = ServerStatus(
+            canBackoff = statusCode in 500..599,
+            statusCode = statusCode,
+            timestamp = now,
+        )
+    }
+
+    fun extractServerFromUrl(url: String): String? {
+        return serverPattern.find(url)?.groups?.get(1)?.value
+    }
+
+    fun replaceServerInUrl(url: String, newServer: String): String {
+        return url.replace(serverPattern, "https://$newServer")
     }
 }

--- a/src/all/batoto/src/eu/kanade/tachiyomi/extension/all/batoto/BatoTo.kt
+++ b/src/all/batoto/src/eu/kanade/tachiyomi/extension/all/batoto/BatoTo.kt
@@ -203,9 +203,8 @@ class ImageServerManager() {
 
     val fallbackServers = listOf(
         "n03", "n00", "n01", "n02", "n04", "n05", "n06", "n07", "n08", "n09", "n10",
-        "k03", "k06", "k00", "k01", "k02", "k04", "k05", "k08", "k09",
     )
-    val blacklist = listOf("k07")
+    val blacklist = listOf("k00", "k01", "k03", "k04", "k05", "k06", "k07", "k08", "k09")
 
     // Server status tracking
     data class ServerStatus(

--- a/src/all/batoto/src/eu/kanade/tachiyomi/extension/all/batotov2/BatoToV2.kt
+++ b/src/all/batoto/src/eu/kanade/tachiyomi/extension/all/batotov2/BatoToV2.kt
@@ -8,6 +8,7 @@ import android.widget.Toast
 import androidx.preference.CheckBoxPreference
 import androidx.preference.EditTextPreference
 import androidx.preference.PreferenceScreen
+import eu.kanade.tachiyomi.extension.all.batoto.ImageServerManager
 import eu.kanade.tachiyomi.lib.cryptoaes.CryptoAES
 import eu.kanade.tachiyomi.lib.cryptoaes.Deobfuscator
 import eu.kanade.tachiyomi.network.GET
@@ -43,7 +44,6 @@ import java.text.ParseException
 import java.text.SimpleDateFormat
 import java.util.Calendar
 import java.util.Locale
-import java.util.concurrent.ConcurrentHashMap
 import java.util.concurrent.TimeUnit
 
 open class BatoToV2(
@@ -1138,7 +1138,6 @@ open class BatoToV2(
     }
 
     companion object {
-        private val SERVER_PATTERN = Regex("https://[a-zA-Z]\\d{2}")
         private val seriesUrlRegex = Regex(""".*/series/(\d+)/.*""")
         private val seriesIdRegex = Regex("""series/(\d+)""")
         private val chapterIdRegex = Regex("""/chapter/(\d+)""") // /chapter/4016325
@@ -1157,55 +1156,5 @@ open class BatoToV2(
 
         private val titleRegex: Regex =
             Regex("\\([^()]*\\)|\\{[^{}]*\\}|\\[(?:(?!]).)*]|«[^»]*»|〘[^〙]*〙|「[^」]*」|『[^』]*』|≪[^≫]*≫|﹛[^﹜]*﹜|〖[^〖〗]*〗|\uD81A\uDD0D.+?\uD81A\uDD0D|《[^》]*》|⌜.+?⌝|⟨[^⟩]*⟩|/Official|/ Official", RegexOption.IGNORE_CASE)
-    }
-}
-
-/**
- * Manages image server fallback logic, including blacklisting and backoff tracking.
- */
-private class ImageServerManager() {
-    val serverPattern = Regex("https://([a-zA-Z]\\d{2})")
-
-    val fallbackServers = listOf(
-        "n03", "n00", "n01", "n02", "n04", "n05", "n06", "n07", "n08", "n09", "n10",
-        "k03", "k06", "k00", "k01", "k02", "k04", "k05", "k08", "k09",
-    )
-    val blacklist = listOf("k07")
-
-    // Server status tracking
-    data class ServerStatus(
-        val canBackoff: Boolean,
-        val statusCode: Int = 0,
-        val timestamp: Long = System.currentTimeMillis(),
-    )
-
-    private val serverStatus = ConcurrentHashMap<String, ServerStatus>()
-
-    private val BACKOFF_DURATION_MS = 3_600_000L // 1 hour
-
-    fun shouldSkip(server: String): Boolean {
-        return server in blacklist || isInBackoff(server)
-    }
-
-    fun isInBackoff(server: String): Boolean {
-        val status = serverStatus[server] ?: return false
-        return status.canBackoff && System.currentTimeMillis() - status.timestamp < BACKOFF_DURATION_MS
-    }
-
-    fun recordImageServerStatus(server: String, statusCode: Int) {
-        val now = System.currentTimeMillis()
-        serverStatus[server] = ServerStatus(
-            canBackoff = statusCode in 500..599,
-            statusCode = statusCode,
-            timestamp = now,
-        )
-    }
-
-    fun extractServerFromUrl(url: String): String? {
-        return serverPattern.find(url)?.groups?.get(1)?.value
-    }
-
-    fun replaceServerInUrl(url: String, newServer: String): String {
-        return url.replace(serverPattern, "https://$newServer")
     }
 }

--- a/src/all/batoto/src/eu/kanade/tachiyomi/extension/all/batotov2/BatoToV2.kt
+++ b/src/all/batoto/src/eu/kanade/tachiyomi/extension/all/batotov2/BatoToV2.kt
@@ -43,6 +43,7 @@ import java.text.ParseException
 import java.text.SimpleDateFormat
 import java.util.Calendar
 import java.util.Locale
+import java.util.concurrent.ConcurrentHashMap
 import java.util.concurrent.TimeUnit
 
 open class BatoToV2(
@@ -66,6 +67,8 @@ open class BatoToV2(
             chain.proceed(request)
         }
         .build()
+
+    private val imageServerManager: ImageServerManager = ImageServerManager()
 
     private val json: Json by injectLazy()
 
@@ -618,52 +621,56 @@ open class BatoToV2(
 
     private fun imageFallbackInterceptor(chain: Interceptor.Chain): Response {
         val request = chain.request()
-        val response = chain.proceed(request)
 
-        if (response.isSuccessful) return response
-
+        // Extract server, or proceed normally if not found
         val urlString = request.url.toString()
+        val originalServer = this.imageServerManager.extractServerFromUrl(urlString) ?: return chain.proceed(request)
 
-        // We know the first attempt failed. Close the response body to release the
-        // connection from the pool and prevent resource leaks before we start the retry loop.
-        // This is critical; otherwise, new requests in the loop may hang or fail.
-        response.close()
-
-        if (SERVER_PATTERN.containsMatchIn(urlString)) {
-            // Sorted list: Most reliable servers FIRST
-            val servers = listOf("n03", "n00", "n01", "n02", "n04", "n05", "n06", "n07", "n08", "n09", "n10", "k03", "k06", "k07", "k00", "k01", "k02", "k04", "k05", "k08", "k09")
-
-            for (server in servers) {
-                val newUrl = urlString.replace(SERVER_PATTERN, "https://$server")
-
-                // Skip if we are about to try the exact same URL that just failed
-                if (newUrl == urlString) continue
-
-                val newRequest = request.newBuilder()
-                    .url(newUrl)
-                    .build()
-
-                try {
-                    // FORCE SHORT TIMEOUTS FOR FALLBACKS
-                    // If a fallback server doesn't answer in 5 seconds, kill it and move to next.
-                    val newResponse = chain
-                        .withConnectTimeout(5, TimeUnit.SECONDS)
-                        .withReadTimeout(10, TimeUnit.SECONDS)
-                        .proceed(newRequest)
-
-                    if (newResponse.isSuccessful) {
-                        return newResponse
-                    }
-                    // If this server also failed, close and loop to the next one
-                    newResponse.close()
-                } catch (_: Exception) {
-                    // Connection error on this mirror, ignore and loop to next
-                }
-            }
+        // Try original server if not skipped
+        if (!this.imageServerManager.shouldSkip(originalServer)) {
+            tryServer(chain, request, originalServer)?.let { return it }
         }
 
-        // If everything failed, re-run original request to return the standard error
+        // Try fallback servers
+        for (server in this.imageServerManager.fallbackServers) {
+            if (this.imageServerManager.shouldSkip(server)) continue
+
+            val newUrl = this.imageServerManager.replaceServerInUrl(urlString, server)
+            val newRequest = request.newBuilder().url(newUrl).build()
+
+            tryServer(chain, newRequest, server, withTimeout = true)?.let { return it }
+        }
+
         return chain.proceed(request)
+    }
+
+    private fun tryServer(
+        chain: Interceptor.Chain,
+        request: Request,
+        server: String,
+        withTimeout: Boolean = false,
+    ): Response? {
+        return try {
+            val modifiedChain = if (withTimeout) {
+                chain.withConnectTimeout(5, TimeUnit.SECONDS)
+                    .withReadTimeout(10, TimeUnit.SECONDS)
+            } else {
+                chain
+            }
+
+            val response = modifiedChain.proceed(request)
+            imageServerManager.recordImageServerStatus(server, response.code)
+
+            if (response.isSuccessful) {
+                response
+            } else {
+                response.close()
+                null
+            }
+        } catch (_: Exception) {
+            imageServerManager.recordImageServerStatus(server, 0) // Unknown error
+            null
+        }
     }
 
     override fun getFilterList() = FilterList(
@@ -1148,5 +1155,55 @@ open class BatoToV2(
 
         private val titleRegex: Regex =
             Regex("\\([^()]*\\)|\\{[^{}]*\\}|\\[(?:(?!]).)*]|«[^»]*»|〘[^〙]*〙|「[^」]*」|『[^』]*』|≪[^≫]*≫|﹛[^﹜]*﹜|〖[^〖〗]*〗|\uD81A\uDD0D.+?\uD81A\uDD0D|《[^》]*》|⌜.+?⌝|⟨[^⟩]*⟩|/Official|/ Official", RegexOption.IGNORE_CASE)
+    }
+}
+
+/**
+ * Manages image server fallback logic, including blacklisting and backoff tracking.
+ */
+private class ImageServerManager() {
+    val serverPattern = Regex("https://([a-zA-Z]\\d{2})")
+
+    val fallbackServers = listOf(
+        "n03", "n00", "n01", "n02", "n04", "n05", "n06", "n07", "n08", "n09", "n10",
+        "k03", "k06", "k00", "k01", "k02", "k04", "k05", "k08", "k09",
+    )
+    val blacklist = listOf("k07")
+
+    // Server status tracking
+    data class ServerStatus(
+        val canBackoff: Boolean,
+        val statusCode: Int = 0,
+        val timestamp: Long = System.currentTimeMillis(),
+    )
+
+    private val serverStatus = ConcurrentHashMap<String, ServerStatus>()
+
+    private val BACKOFF_DURATION_MS = 3_600_000L // 1 hour
+
+    fun shouldSkip(server: String): Boolean {
+        return server in blacklist || isInBackoff(server)
+    }
+
+    fun isInBackoff(server: String): Boolean {
+        val status = serverStatus[server] ?: return false
+        return status.canBackoff && System.currentTimeMillis() - status.timestamp < BACKOFF_DURATION_MS
+    }
+
+    fun recordImageServerStatus(server: String, statusCode: Int) {
+        val now = System.currentTimeMillis()
+        serverStatus[server] = ServerStatus(
+            canBackoff = statusCode in 500..599,
+            statusCode = statusCode,
+            timestamp = now,
+        )
+    }
+
+    fun extractServerFromUrl(url: String): String? {
+        return serverPattern.find(url)?.groups?.get(1)?.value
+    }
+
+    fun replaceServerInUrl(url: String, newServer: String): String {
+        return url.replace(serverPattern, "https://$newServer")
     }
 }

--- a/src/all/batoto/src/eu/kanade/tachiyomi/extension/all/batotov2/BatoToV2.kt
+++ b/src/all/batoto/src/eu/kanade/tachiyomi/extension/all/batotov2/BatoToV2.kt
@@ -651,6 +651,8 @@ open class BatoToV2(
         withTimeout: Boolean = false,
     ): Response? {
         return try {
+            // FORCE SHORT TIMEOUTS FOR FALLBACKS
+            // If a fallback server doesn't answer in 5 seconds, kill it and move to next server.
             val modifiedChain = if (withTimeout) {
                 chain.withConnectTimeout(5, TimeUnit.SECONDS)
                     .withReadTimeout(10, TimeUnit.SECONDS)

--- a/src/all/batoto/src/eu/kanade/tachiyomi/extension/all/batotov4/BatoToV4.kt
+++ b/src/all/batoto/src/eu/kanade/tachiyomi/extension/all/batotov4/BatoToV4.kt
@@ -8,6 +8,7 @@ import android.widget.Toast
 import androidx.preference.CheckBoxPreference
 import androidx.preference.EditTextPreference
 import androidx.preference.PreferenceScreen
+import eu.kanade.tachiyomi.extension.all.batoto.ImageServerManager
 import eu.kanade.tachiyomi.network.GET
 import eu.kanade.tachiyomi.network.POST
 import eu.kanade.tachiyomi.network.asObservableSuccess
@@ -34,7 +35,6 @@ import okhttp3.Response
 import okhttp3.internal.closeQuietly
 import okio.IOException
 import rx.Observable
-import java.util.concurrent.ConcurrentHashMap
 import java.util.concurrent.TimeUnit
 
 class BatoToV4(
@@ -577,53 +577,3 @@ private const val BROWSE_PAGE_SIZE = 36
 private val seriesIdRegex = Regex("""series/(\d+)""")
 private val titleRegex: Regex =
     Regex("\\([^()]*\\)|\\{[^{}]*\\}|\\[(?:(?!]).)*]|«[^»]*»|〘[^〙]*〙|「[^」]*」|『[^』]*』|≪[^≫]*≫|﹛[^﹜]*﹜|〖[^〖〗]*〗|\uD81A\uDD0D.+?\uD81A\uDD0D|《[^》]*》|⌜.+?⌝|⟨[^⟩]*⟩|/Official|/ Official", RegexOption.IGNORE_CASE)
-
-/**
- * Manages image server fallback logic, including blacklisting and backoff tracking.
- */
-private class ImageServerManager() {
-    val serverPattern = Regex("https://([a-zA-Z]\\d{2})")
-
-    val fallbackServers = listOf(
-        "n03", "n00", "n01", "n02", "n04", "n05", "n06", "n07", "n08", "n09", "n10",
-        "k03", "k06", "k00", "k01", "k02", "k04", "k05", "k08", "k09",
-    )
-    val blacklist = listOf("k07")
-
-    // Server status tracking
-    data class ServerStatus(
-        val canBackoff: Boolean,
-        val statusCode: Int = 0,
-        val timestamp: Long = System.currentTimeMillis(),
-    )
-
-    private val serverStatus = ConcurrentHashMap<String, ServerStatus>()
-
-    private val BACKOFF_DURATION_MS = 3_600_000L // 1 hour
-
-    fun shouldSkip(server: String): Boolean {
-        return server in blacklist || isInBackoff(server)
-    }
-
-    fun isInBackoff(server: String): Boolean {
-        val status = serverStatus[server] ?: return false
-        return status.canBackoff && System.currentTimeMillis() - status.timestamp < BACKOFF_DURATION_MS
-    }
-
-    fun recordImageServerStatus(server: String, statusCode: Int) {
-        val now = System.currentTimeMillis()
-        serverStatus[server] = ServerStatus(
-            canBackoff = statusCode in 500..599,
-            statusCode = statusCode,
-            timestamp = now,
-        )
-    }
-
-    fun extractServerFromUrl(url: String): String? {
-        return serverPattern.find(url)?.groups?.get(1)?.value
-    }
-
-    fun replaceServerInUrl(url: String, newServer: String): String {
-        return url.replace(serverPattern, "https://$newServer")
-    }
-}

--- a/src/all/batoto/src/eu/kanade/tachiyomi/extension/all/batotov4/BatoToV4.kt
+++ b/src/all/batoto/src/eu/kanade/tachiyomi/extension/all/batotov4/BatoToV4.kt
@@ -572,9 +572,7 @@ class BatoToV4(
 
             setOnPreferenceChangeListener { _, newValue ->
                 val blacklistStr = newValue as String
-                val blacklist = blacklistStr.split(",")
-                    .map { it.trim() }
-                    .filter { it.isNotEmpty() }
+                val blacklist = parseCommaList(blacklistStr)
 
                 imageServerManager.updateBlacklist(blacklist)
 
@@ -594,13 +592,7 @@ class BatoToV4(
 
             setOnPreferenceChangeListener { _, newValue ->
                 val fallbackStr = newValue as String
-                val fallbackList = if (fallbackStr.isEmpty()) {
-                    null
-                } else {
-                    fallbackStr.split(",")
-                        .map { it.trim() }
-                        .filter { it.isNotEmpty() }
-                }
+                val fallbackList = parseCommaList(fallbackStr)
 
                 imageServerManager.updateFallbackServers(fallbackList)
 
@@ -625,15 +617,9 @@ class BatoToV4(
 
     private fun getBlacklistSummary(): String {
         val blacklistStr = preferences.getString(BLACKLIST_SERVERS_PREF, "")!!
-        val blacklist = if (blacklistStr.isEmpty()) {
-            imageServerManager.getDefaultBlacklistedServers()
-        } else {
-            blacklistStr.split(",")
-                .map { it.trim() }
-                .filter { it.isNotEmpty() }
-        }
-        return if (blacklist.isEmpty()) {
-            "Using default: ${imageServerManager.getDefaultBlacklistedServers().joinToString(", ")}"
+        val blacklist = parseCommaList(blacklistStr) ?: imageServerManager.getDefaultBlacklistedServers()
+        return if (blacklist == imageServerManager.getDefaultBlacklistedServers()) {
+            "Using default: ${blacklist.joinToString(", ")}"
         } else {
             "Blacklisted: ${blacklist.joinToString(", ")}"
         }
@@ -641,15 +627,9 @@ class BatoToV4(
 
     private fun getFallbackSummary(fallbackStr: String? = null): String {
         val fallback = fallbackStr ?: preferences.getString(FALLBACK_SERVERS_PREF, "")!!
-        val servers = if (fallback.isEmpty()) {
-            imageServerManager.getDefaultFallbackServers()
-        } else {
-            fallback.split(",")
-                .map { it.trim() }
-                .filter { it.isNotEmpty() }
-        }
-        return if (servers.isEmpty()) {
-            "Using default: ${imageServerManager.getDefaultFallbackServers().joinToString(", ")}"
+        val servers = parseCommaList(fallback) ?: imageServerManager.getDefaultFallbackServers()
+        return if (servers == imageServerManager.getDefaultFallbackServers()) {
+            "Using default: ${servers.joinToString(", ")}"
         } else {
             "Fallback order: ${servers.joinToString(", ")}"
         }
@@ -686,6 +666,14 @@ class BatoToV4(
         }
 
         return explanation + serverList
+    }
+
+    private fun parseCommaList(value: String?): List<String>? {
+        if (value.isNullOrEmpty()) return null
+        return value.split(",")
+            .map { it.trim() }
+            .filter { it.isNotEmpty() }
+            .takeIf { it.isNotEmpty() }
     }
 
     private fun isRemoveTitleVersion(): Boolean {

--- a/src/all/batoto/src/eu/kanade/tachiyomi/extension/all/batotov4/BatoToV4.kt
+++ b/src/all/batoto/src/eu/kanade/tachiyomi/extension/all/batotov4/BatoToV4.kt
@@ -34,6 +34,7 @@ import okhttp3.Response
 import okhttp3.internal.closeQuietly
 import okio.IOException
 import rx.Observable
+import java.util.concurrent.ConcurrentHashMap
 import java.util.concurrent.TimeUnit
 
 class BatoToV4(
@@ -44,6 +45,8 @@ class BatoToV4(
 ) : ConfigurableSource, HttpSource() {
 
     override val name: String = "Bato.to"
+
+    private val imageServerManager: ImageServerManager = ImageServerManager()
 
     override val supportsLatest = true
 
@@ -413,45 +416,63 @@ class BatoToV4(
 
     private fun imageFallbackInterceptor(chain: Interceptor.Chain): Response {
         val request = chain.request()
-        val response = chain.proceed(request)
 
-        if (request.url.fragment != PAGE_FRAGMENT || response.isSuccessful) {
-            return response
+        // Handle non-page requests normally
+        if (request.url.fragment != PAGE_FRAGMENT) {
+            return chain.proceed(request)
         }
 
-        response.closeQuietly()
-
+        // Extract server, or proceed normally if not found
         val urlString = request.url.toString()
+        val originalServer = this.imageServerManager.extractServerFromUrl(urlString) ?: return chain.proceed(request)
 
-        if (SERVER_PATTERN.containsMatchIn(urlString)) {
-            // Sorted list: Most reliable servers FIRST
-            val servers = listOf("n03", "n00", "n01", "n02", "n04", "n05", "n06", "n07", "n08", "n09", "n10", "k03", "k06", "k07", "k00", "k01", "k02", "k04", "k05", "k08", "k09")
+        val backoffEnabled = preferences.getBoolean(ENABLE_BACKOFF_PREF, true)
 
-            for (server in servers) {
-                val newUrl = urlString.replace(SERVER_PATTERN, "https://$server")
+        // Try original server if not skipped
+        if (!this.imageServerManager.shouldSkip(originalServer, backoffEnabled)) {
+            tryServer(chain, request, originalServer)?.let { return it }
+        }
 
-                val newRequest = request.newBuilder()
-                    .url(newUrl)
-                    .build()
+        // Try fallback servers
+        for (server in this.imageServerManager.getValidServers()) {
+            if (this.imageServerManager.shouldSkip(server, backoffEnabled)) continue
 
-                try {
-                    val newResponse = chain
-                        .withConnectTimeout(5, TimeUnit.SECONDS)
-                        .withReadTimeout(10, TimeUnit.SECONDS)
-                        .proceed(newRequest)
+            val newUrl = this.imageServerManager.replaceServerInUrl(urlString, server)
+            val newRequest = request.newBuilder().url(newUrl).build()
 
-                    if (newResponse.isSuccessful) {
-                        return newResponse
-                    }
-
-                    newResponse.close()
-                } catch (_: Exception) {
-                    // Connection error on this mirror, ignore and loop to next
-                }
-            }
+            tryServer(chain, newRequest, server, withTimeout = true)?.let { return it }
         }
 
         return chain.proceed(request)
+    }
+
+    private fun tryServer(
+        chain: Interceptor.Chain,
+        request: Request,
+        server: String,
+        withTimeout: Boolean = false,
+    ): Response? {
+        return try {
+            val modifiedChain = if (withTimeout) {
+                chain.withConnectTimeout(5, TimeUnit.SECONDS)
+                    .withReadTimeout(10, TimeUnit.SECONDS)
+            } else {
+                chain
+            }
+
+            val response = modifiedChain.proceed(request)
+            imageServerManager.recordImageServerStatus(server, response.code)
+
+            if (response.isSuccessful) {
+                response
+            } else {
+                response.close()
+                null
+            }
+        } catch (_: Exception) {
+            imageServerManager.recordImageServerStatus(server, 0) // Unknown error
+            null
+        }
     }
 
     override fun getFilterList(): FilterList {
@@ -539,6 +560,132 @@ class BatoToV4(
                 isValid
             }
         }.also(screen::addPreference)
+
+        // Blacklisted image servers preference
+        EditTextPreference(screen.context).apply {
+            key = BLACKLIST_SERVERS_PREF
+            title = "Blacklisted image servers"
+            summary = getBlacklistSummary()
+            setDefaultValue("")
+            dialogTitle = "Blacklisted servers"
+            dialogMessage = "Comma-separated list of server names to skip (e.g. k07,k01,n03). Leave empty to use default."
+
+            setOnPreferenceChangeListener { _, newValue ->
+                val blacklistStr = newValue as String
+                val blacklist = blacklistStr.split(",")
+                    .map { it.trim() }
+                    .filter { it.isNotEmpty() }
+
+                imageServerManager.updateBlacklist(blacklist)
+
+                summary = getBlacklistSummary()
+                true
+            }
+        }.also(screen::addPreference)
+
+        // Fallback servers list preference
+        EditTextPreference(screen.context).apply {
+            key = FALLBACK_SERVERS_PREF
+            title = "Fallback image servers"
+            summary = getFallbackSummary()
+            setDefaultValue("")
+            dialogTitle = "Fallback servers"
+            dialogMessage = "Comma-separated list of server names in order of preference (e.g. n03,n00,k06). Leave empty to use default."
+
+            setOnPreferenceChangeListener { _, newValue ->
+                val fallbackStr = newValue as String
+                val fallbackList = if (fallbackStr.isEmpty()) {
+                    null
+                } else {
+                    fallbackStr.split(",")
+                        .map { it.trim() }
+                        .filter { it.isNotEmpty() }
+                }
+
+                imageServerManager.updateFallbackServers(fallbackList)
+
+                summary = getFallbackSummary(fallbackStr)
+                true
+            }
+        }.also(screen::addPreference)
+
+        // Skip failed servers (backoff) preference
+        CheckBoxPreference(screen.context).apply {
+            key = ENABLE_BACKOFF_PREF
+            title = "Skip failed servers"
+            summary = getBackoffSummary()
+            setDefaultValue(true)
+
+            setOnPreferenceChangeListener { _, newValue ->
+                summary = getBackoffSummary(newValue as Boolean)
+                true
+            }
+        }.also(screen::addPreference)
+    }
+
+    private fun getBlacklistSummary(): String {
+        val blacklistStr = preferences.getString(BLACKLIST_SERVERS_PREF, "")!!
+        val blacklist = if (blacklistStr.isEmpty()) {
+            imageServerManager.getDefaultBlacklistedServers()
+        } else {
+            blacklistStr.split(",")
+                .map { it.trim() }
+                .filter { it.isNotEmpty() }
+        }
+        return if (blacklist.isEmpty()) {
+            "Using default: ${imageServerManager.getDefaultBlacklistedServers().joinToString(", ")}"
+        } else {
+            "Blacklisted: ${blacklist.joinToString(", ")}"
+        }
+    }
+
+    private fun getFallbackSummary(fallbackStr: String? = null): String {
+        val fallback = fallbackStr ?: preferences.getString(FALLBACK_SERVERS_PREF, "")!!
+        val servers = if (fallback.isEmpty()) {
+            imageServerManager.getDefaultFallbackServers()
+        } else {
+            fallback.split(",")
+                .map { it.trim() }
+                .filter { it.isNotEmpty() }
+        }
+        return if (servers.isEmpty()) {
+            "Using default: ${imageServerManager.getDefaultFallbackServers().joinToString(", ")}"
+        } else {
+            "Fallback order: ${servers.joinToString(", ")}"
+        }
+    }
+
+    private fun getBackoffSummary(isEnabled: Boolean? = null): String {
+        val serverStatuses = imageServerManager.getAllServerStatuses()
+        val enabled = isEnabled ?: preferences.getBoolean(ENABLE_BACKOFF_PREF, true)
+
+        val explanation = if (enabled) "Automatically skipping servers that return 500 errors for 1 hour" else "Not skipping servers that return 500 errors"
+
+        val inBackoffServers = serverStatuses.entries.filter { imageServerManager.isInBackoff(it.key) }
+
+        val serverList = if (enabled) {
+            if (inBackoffServers.isEmpty()) {
+                "\nNo servers in backoff"
+            } else {
+                "\nServers in backoff:\n" + inBackoffServers
+                    .sortedBy { it.key }
+                    .joinToString("\n") { (server, _) ->
+                        "$server: ${imageServerManager.getBackoffTimeLeftString(server)}"
+                    }
+            }
+        } else {
+            if (inBackoffServers.isEmpty()) {
+                "\nNo servers recently failed"
+            } else {
+                "\nServers recently failed:\n" + inBackoffServers
+                    .sortedBy { it.key }
+                    .joinToString("\n") { (server, status) ->
+                        "$server: error ${status.statusCode} (${imageServerManager.getTimeSinceStatus(server)} ago)"
+                    }
+            }
+        }
+
+        return explanation + serverList
     }
 
     private fun isRemoveTitleVersion(): Boolean {
@@ -553,10 +700,109 @@ private val jsonMediaType = "application/json".toMediaType()
 
 private const val REMOVE_TITLE_VERSION_PREF = "REMOVE_TITLE_VERSION"
 private const val REMOVE_TITLE_CUSTOM_PREF = "REMOVE_TITLE_CUSTOM"
+private const val BLACKLIST_SERVERS_PREF = "BLACKLIST_SERVERS"
+private const val FALLBACK_SERVERS_PREF = "FALLBACK_SERVERS"
+private const val ENABLE_BACKOFF_PREF = "ENABLE_BACKOFF"
 
 private const val PAGE_FRAGMENT = "page"
-private val SERVER_PATTERN = Regex("https://[a-zA-Z]\\d{2}")
 private val userIdRegex = Regex("""/u/(\d+)""")
+
+/**
+ * Manages image server fallback logic, including blacklisting and backoff tracking.
+ */
+private class ImageServerManager(
+    blacklistedServers: List<String>? = null,
+    fallbackServers: List<String>? = null,
+) {
+    val serverPattern = Regex("https://([a-zA-Z]\\d{2})")
+
+    // Sorted list: Most reliable servers FIRST
+    private val defaultFallbackServers = listOf(
+        "n03", "n00", "n01", "n02", "n04", "n05", "n06", "n07", "n08", "n09", "n10",
+        "k03", "k06", "k00", "k01", "k02", "k04", "k05", "k08", "k09",
+    )
+    private val defaultBlacklistedServers = listOf("k07")
+
+    private var validServers = (fallbackServers ?: defaultFallbackServers).toMutableList()
+    private val blacklist = (blacklistedServers ?: defaultBlacklistedServers).toMutableList()
+
+    // Server status tracking
+    data class ServerStatus(
+        val canBackoff: Boolean,
+        val statusCode: Int = 0,
+        val timestamp: Long = System.currentTimeMillis(),
+    )
+
+    private val serverStatus = ConcurrentHashMap<String, ServerStatus>()
+
+    private val BACKOFF_DURATION_MS = 3_600_000L // 1 hour
+
+    fun shouldSkip(server: String, backoffEnabled: Boolean = true): Boolean {
+        return server in blacklist || (backoffEnabled && isInBackoff(server))
+    }
+
+    fun isInBackoff(server: String): Boolean {
+        val status = serverStatus[server] ?: return false
+        return status.canBackoff && System.currentTimeMillis() - status.timestamp < BACKOFF_DURATION_MS
+    }
+
+    fun recordImageServerStatus(server: String, statusCode: Int) {
+        val now = System.currentTimeMillis()
+        serverStatus[server] = ServerStatus(
+            canBackoff = statusCode in 500..599,
+            statusCode = statusCode,
+            timestamp = now,
+        )
+    }
+
+    fun getBackoffTimeLeftString(server: String): String {
+        val status = serverStatus[server] ?: return "0s"
+        val backoffEnd = status.timestamp + BACKOFF_DURATION_MS
+        val remaining = maxOf(backoffEnd - System.currentTimeMillis(), 0)
+
+        val seconds = remaining / 1000
+        val minutes = seconds / 60
+        val remainingSeconds = seconds % 60
+        return if (minutes > 0) "${minutes}m ${remainingSeconds}s" else "${remainingSeconds}s"
+    }
+
+    fun getTimeSinceStatus(server: String): String {
+        val status = serverStatus[server] ?: return "0s"
+        val diff = System.currentTimeMillis() - status.timestamp
+
+        val seconds = diff / 1000
+        val minutes = seconds / 60
+        val remainingSeconds = seconds % 60
+        return if (minutes > 0) "${minutes}m ${remainingSeconds}s" else "${remainingSeconds}s"
+    }
+
+    fun extractServerFromUrl(url: String): String? {
+        return serverPattern.find(url)?.groups?.get(1)?.value
+    }
+
+    fun replaceServerInUrl(url: String, newServer: String): String {
+        return url.replace(serverPattern, "https://$newServer")
+    }
+
+    fun getAllServerStatuses(): Map<String, ServerStatus> {
+        return serverStatus.toMap()
+    }
+
+    fun getValidServers(): List<String> = validServers.toList()
+
+    fun updateBlacklist(blacklistedServers: List<String>?) {
+        blacklist.clear()
+        blacklist.addAll(blacklistedServers ?: defaultBlacklistedServers)
+    }
+
+    fun updateFallbackServers(servers: List<String>?) {
+        validServers.clear()
+        validServers.addAll(servers ?: defaultFallbackServers)
+    }
+
+    fun getDefaultFallbackServers(): List<String> = defaultFallbackServers
+    fun getDefaultBlacklistedServers(): List<String> = defaultBlacklistedServers
+}
 
 private const val BROWSE_PAGE_SIZE = 36
 

--- a/src/all/batoto/src/eu/kanade/tachiyomi/extension/all/batotov4/BatoToV4.kt
+++ b/src/all/batoto/src/eu/kanade/tachiyomi/extension/all/batotov4/BatoToV4.kt
@@ -46,8 +46,6 @@ class BatoToV4(
 
     override val name: String = "Bato.to"
 
-    private val imageServerManager: ImageServerManager = ImageServerManager()
-
     override val supportsLatest = true
 
     override val client = network.cloudflareClient.newBuilder()
@@ -60,6 +58,8 @@ class BatoToV4(
             chain.proceed(request)
         }
         .build()
+
+    private val imageServerManager: ImageServerManager = ImageServerManager()
 
     // ************ Search ************ //
     override fun latestUpdatesRequest(page: Int) = throw UnsupportedOperationException()
@@ -368,7 +368,6 @@ class BatoToV4(
                 Page(
                     index = index,
                     imageUrl = url.toHttpUrl().newBuilder()
-                        .fragment(PAGE_FRAGMENT)
                         .build()
                         .toString(),
                 )
@@ -416,11 +415,6 @@ class BatoToV4(
 
     private fun imageFallbackInterceptor(chain: Interceptor.Chain): Response {
         val request = chain.request()
-
-        // Handle non-page requests normally
-        if (request.url.fragment != PAGE_FRAGMENT) {
-            return chain.proceed(request)
-        }
 
         // Extract server, or proceed normally if not found
         val urlString = request.url.toString()
@@ -573,7 +567,6 @@ private val jsonMediaType = "application/json".toMediaType()
 private const val REMOVE_TITLE_VERSION_PREF = "REMOVE_TITLE_VERSION"
 private const val REMOVE_TITLE_CUSTOM_PREF = "REMOVE_TITLE_CUSTOM"
 
-private const val PAGE_FRAGMENT = "page"
 private val userIdRegex = Regex("""/u/(\d+)""")
 
 private const val BROWSE_PAGE_SIZE = 36

--- a/src/all/batoto/src/eu/kanade/tachiyomi/extension/all/batotov4/BatoToV4.kt
+++ b/src/all/batoto/src/eu/kanade/tachiyomi/extension/all/batotov4/BatoToV4.kt
@@ -445,6 +445,8 @@ class BatoToV4(
         withTimeout: Boolean = false,
     ): Response? {
         return try {
+            // FORCE SHORT TIMEOUTS FOR FALLBACKS
+            // If a fallback server doesn't answer in 5 seconds, kill it and move to next server.
             val modifiedChain = if (withTimeout) {
                 chain.withConnectTimeout(5, TimeUnit.SECONDS)
                     .withReadTimeout(10, TimeUnit.SECONDS)

--- a/src/all/batoto/src/eu/kanade/tachiyomi/extension/all/batotov4/BatoToV4.kt
+++ b/src/all/batoto/src/eu/kanade/tachiyomi/extension/all/batotov4/BatoToV4.kt
@@ -426,16 +426,14 @@ class BatoToV4(
         val urlString = request.url.toString()
         val originalServer = this.imageServerManager.extractServerFromUrl(urlString) ?: return chain.proceed(request)
 
-        val backoffEnabled = preferences.getBoolean(ENABLE_BACKOFF_PREF, true)
-
         // Try original server if not skipped
-        if (!this.imageServerManager.shouldSkip(originalServer, backoffEnabled)) {
+        if (!this.imageServerManager.shouldSkip(originalServer)) {
             tryServer(chain, request, originalServer)?.let { return it }
         }
 
         // Try fallback servers
-        for (server in this.imageServerManager.getValidServers()) {
-            if (this.imageServerManager.shouldSkip(server, backoffEnabled)) continue
+        for (server in this.imageServerManager.fallbackServers) {
+            if (this.imageServerManager.shouldSkip(server)) continue
 
             val newUrl = this.imageServerManager.replaceServerInUrl(urlString, server)
             val newRequest = request.newBuilder().url(newUrl).build()
@@ -560,120 +558,6 @@ class BatoToV4(
                 isValid
             }
         }.also(screen::addPreference)
-
-        // Blacklisted image servers preference
-        EditTextPreference(screen.context).apply {
-            key = BLACKLIST_SERVERS_PREF
-            title = "Blacklisted image servers"
-            summary = getBlacklistSummary()
-            setDefaultValue("")
-            dialogTitle = "Blacklisted servers"
-            dialogMessage = "Comma-separated list of server names to skip (e.g. k07,k01,n03). Leave empty to use default."
-
-            setOnPreferenceChangeListener { _, newValue ->
-                val blacklistStr = newValue as String
-                val blacklist = parseCommaList(blacklistStr)
-
-                imageServerManager.updateBlacklist(blacklist)
-
-                summary = getBlacklistSummary()
-                true
-            }
-        }.also(screen::addPreference)
-
-        // Fallback servers list preference
-        EditTextPreference(screen.context).apply {
-            key = FALLBACK_SERVERS_PREF
-            title = "Fallback image servers"
-            summary = getFallbackSummary()
-            setDefaultValue("")
-            dialogTitle = "Fallback servers"
-            dialogMessage = "Comma-separated list of server names in order of preference (e.g. n03,n00,k06). Leave empty to use default."
-
-            setOnPreferenceChangeListener { _, newValue ->
-                val fallbackStr = newValue as String
-                val fallbackList = parseCommaList(fallbackStr)
-
-                imageServerManager.updateFallbackServers(fallbackList)
-
-                summary = getFallbackSummary(fallbackStr)
-                true
-            }
-        }.also(screen::addPreference)
-
-        // Skip failed servers (backoff) preference
-        CheckBoxPreference(screen.context).apply {
-            key = ENABLE_BACKOFF_PREF
-            title = "Skip failed servers"
-            summary = getBackoffSummary()
-            setDefaultValue(true)
-
-            setOnPreferenceChangeListener { _, newValue ->
-                summary = getBackoffSummary(newValue as Boolean)
-                true
-            }
-        }.also(screen::addPreference)
-    }
-
-    private fun getBlacklistSummary(): String {
-        val blacklistStr = preferences.getString(BLACKLIST_SERVERS_PREF, "")!!
-        val blacklist = parseCommaList(blacklistStr) ?: imageServerManager.getDefaultBlacklistedServers()
-        return if (blacklist == imageServerManager.getDefaultBlacklistedServers()) {
-            "Using default: ${blacklist.joinToString(", ")}"
-        } else {
-            "Blacklisted: ${blacklist.joinToString(", ")}"
-        }
-    }
-
-    private fun getFallbackSummary(fallbackStr: String? = null): String {
-        val fallback = fallbackStr ?: preferences.getString(FALLBACK_SERVERS_PREF, "")!!
-        val servers = parseCommaList(fallback) ?: imageServerManager.getDefaultFallbackServers()
-        return if (servers == imageServerManager.getDefaultFallbackServers()) {
-            "Using default: ${servers.joinToString(", ")}"
-        } else {
-            "Fallback order: ${servers.joinToString(", ")}"
-        }
-    }
-
-    private fun getBackoffSummary(isEnabled: Boolean? = null): String {
-        val serverStatuses = imageServerManager.getAllServerStatuses()
-        val enabled = isEnabled ?: preferences.getBoolean(ENABLE_BACKOFF_PREF, true)
-
-        val explanation = if (enabled) "Automatically skipping servers that return 500 errors for 1 hour" else "Not skipping servers that return 500 errors"
-
-        val inBackoffServers = serverStatuses.entries.filter { imageServerManager.isInBackoff(it.key) }
-
-        val serverList = if (enabled) {
-            if (inBackoffServers.isEmpty()) {
-                "\nNo servers in backoff"
-            } else {
-                "\nServers in backoff:\n" + inBackoffServers
-                    .sortedBy { it.key }
-                    .joinToString("\n") { (server, _) ->
-                        "$server: ${imageServerManager.getBackoffTimeLeftString(server)}"
-                    }
-            }
-        } else {
-            if (inBackoffServers.isEmpty()) {
-                "\nNo servers recently failed"
-            } else {
-                "\nServers recently failed:\n" + inBackoffServers
-                    .sortedBy { it.key }
-                    .joinToString("\n") { (server, status) ->
-                        "$server: error ${status.statusCode} (${imageServerManager.getTimeSinceStatus(server)} ago)"
-                    }
-            }
-        }
-
-        return explanation + serverList
-    }
-
-    private fun parseCommaList(value: String?): List<String>? {
-        if (value.isNullOrEmpty()) return null
-        return value.split(",")
-            .map { it.trim() }
-            .filter { it.isNotEmpty() }
-            .takeIf { it.isNotEmpty() }
     }
 
     private fun isRemoveTitleVersion(): Boolean {
@@ -688,9 +572,6 @@ private val jsonMediaType = "application/json".toMediaType()
 
 private const val REMOVE_TITLE_VERSION_PREF = "REMOVE_TITLE_VERSION"
 private const val REMOVE_TITLE_CUSTOM_PREF = "REMOVE_TITLE_CUSTOM"
-private const val BLACKLIST_SERVERS_PREF = "BLACKLIST_SERVERS"
-private const val FALLBACK_SERVERS_PREF = "FALLBACK_SERVERS"
-private const val ENABLE_BACKOFF_PREF = "ENABLE_BACKOFF"
 
 private const val PAGE_FRAGMENT = "page"
 private val userIdRegex = Regex("""/u/(\d+)""")
@@ -698,21 +579,14 @@ private val userIdRegex = Regex("""/u/(\d+)""")
 /**
  * Manages image server fallback logic, including blacklisting and backoff tracking.
  */
-private class ImageServerManager(
-    blacklistedServers: List<String>? = null,
-    fallbackServers: List<String>? = null,
-) {
+private class ImageServerManager() {
     val serverPattern = Regex("https://([a-zA-Z]\\d{2})")
 
-    // Sorted list: Most reliable servers FIRST
-    private val defaultFallbackServers = listOf(
+    val fallbackServers = listOf(
         "n03", "n00", "n01", "n02", "n04", "n05", "n06", "n07", "n08", "n09", "n10",
         "k03", "k06", "k00", "k01", "k02", "k04", "k05", "k08", "k09",
     )
-    private val defaultBlacklistedServers = listOf("k07")
-
-    private var validServers = (fallbackServers ?: defaultFallbackServers).toMutableList()
-    private val blacklist = (blacklistedServers ?: defaultBlacklistedServers).toMutableList()
+    val blacklist = listOf("k07")
 
     // Server status tracking
     data class ServerStatus(
@@ -725,8 +599,8 @@ private class ImageServerManager(
 
     private val BACKOFF_DURATION_MS = 3_600_000L // 1 hour
 
-    fun shouldSkip(server: String, backoffEnabled: Boolean = true): Boolean {
-        return server in blacklist || (backoffEnabled && isInBackoff(server))
+    fun shouldSkip(server: String): Boolean {
+        return server in blacklist || isInBackoff(server)
     }
 
     fun isInBackoff(server: String): Boolean {
@@ -743,27 +617,6 @@ private class ImageServerManager(
         )
     }
 
-    fun getBackoffTimeLeftString(server: String): String {
-        val status = serverStatus[server] ?: return "0s"
-        val backoffEnd = status.timestamp + BACKOFF_DURATION_MS
-        val remaining = maxOf(backoffEnd - System.currentTimeMillis(), 0)
-
-        val seconds = remaining / 1000
-        val minutes = seconds / 60
-        val remainingSeconds = seconds % 60
-        return if (minutes > 0) "${minutes}m ${remainingSeconds}s" else "${remainingSeconds}s"
-    }
-
-    fun getTimeSinceStatus(server: String): String {
-        val status = serverStatus[server] ?: return "0s"
-        val diff = System.currentTimeMillis() - status.timestamp
-
-        val seconds = diff / 1000
-        val minutes = seconds / 60
-        val remainingSeconds = seconds % 60
-        return if (minutes > 0) "${minutes}m ${remainingSeconds}s" else "${remainingSeconds}s"
-    }
-
     fun extractServerFromUrl(url: String): String? {
         return serverPattern.find(url)?.groups?.get(1)?.value
     }
@@ -771,25 +624,6 @@ private class ImageServerManager(
     fun replaceServerInUrl(url: String, newServer: String): String {
         return url.replace(serverPattern, "https://$newServer")
     }
-
-    fun getAllServerStatuses(): Map<String, ServerStatus> {
-        return serverStatus.toMap()
-    }
-
-    fun getValidServers(): List<String> = validServers.toList()
-
-    fun updateBlacklist(blacklistedServers: List<String>?) {
-        blacklist.clear()
-        blacklist.addAll(blacklistedServers ?: defaultBlacklistedServers)
-    }
-
-    fun updateFallbackServers(servers: List<String>?) {
-        validServers.clear()
-        validServers.addAll(servers ?: defaultFallbackServers)
-    }
-
-    fun getDefaultFallbackServers(): List<String> = defaultFallbackServers
-    fun getDefaultBlacklistedServers(): List<String> = defaultBlacklistedServers
 }
 
 private const val BROWSE_PAGE_SIZE = 36

--- a/src/all/batoto/src/eu/kanade/tachiyomi/extension/all/batotov4/BatoToV4.kt
+++ b/src/all/batoto/src/eu/kanade/tachiyomi/extension/all/batotov4/BatoToV4.kt
@@ -576,6 +576,13 @@ private const val REMOVE_TITLE_CUSTOM_PREF = "REMOVE_TITLE_CUSTOM"
 private const val PAGE_FRAGMENT = "page"
 private val userIdRegex = Regex("""/u/(\d+)""")
 
+private const val BROWSE_PAGE_SIZE = 36
+
+// Match old v2 Url
+private val seriesIdRegex = Regex("""series/(\d+)""")
+private val titleRegex: Regex =
+    Regex("\\([^()]*\\)|\\{[^{}]*\\}|\\[(?:(?!]).)*]|«[^»]*»|〘[^〙]*〙|「[^」]*」|『[^』]*』|≪[^≫]*≫|﹛[^﹜]*﹜|〖[^〖〗]*〗|\uD81A\uDD0D.+?\uD81A\uDD0D|《[^》]*》|⌜.+?⌝|⟨[^⟩]*⟩|/Official|/ Official", RegexOption.IGNORE_CASE)
+
 /**
  * Manages image server fallback logic, including blacklisting and backoff tracking.
  */
@@ -625,10 +632,3 @@ private class ImageServerManager() {
         return url.replace(serverPattern, "https://$newServer")
     }
 }
-
-private const val BROWSE_PAGE_SIZE = 36
-
-// Match old v2 Url
-private val seriesIdRegex = Regex("""series/(\d+)""")
-private val titleRegex: Regex =
-    Regex("\\([^()]*\\)|\\{[^{}]*\\}|\\[(?:(?!]).)*]|«[^»]*»|〘[^〙]*〙|「[^」]*」|『[^』]*』|≪[^≫]*≫|﹛[^﹜]*﹜|〖[^〖〗]*〗|\uD81A\uDD0D.+?\uD81A\uDD0D|《[^》]*》|⌜.+?⌝|⟨[^⟩]*⟩|/Official|/ Official", RegexOption.IGNORE_CASE)

